### ISSUE TITLE
3rdParty Support for AMP-STORY

### ISSUE THREAD

**Issue by cmaurersp** [ampproject/amphtml#36793]
### Description

As a 3rd party consent provider/CMP, I want to be able to integrate our TCF version 2 consent solution in AMP-STORY pages. Does this capability currently exist?

### Alternatives Considered

Our consent module for AMP consent does not work in the context of an AMP-STORY page

**Comment by newmuis** [comment]
@cmaurersp can you provide an example of what you've tried?

**Comment by vyoungnyc** [comment]
@newmuis we would like to add support for external consent UI flow via promptUiSrc like in regular amp pages. At the moment from the documentation, amp-story only has support for a default UI prompt. 

seems this was also brought up by someone else in the summer, and doesnt seem to be supported yet:
https://github.com/ampproject/amphtml/issues/35463

**Comment by newmuis** [comment]
Makes sense.  We would need to plan around this feature and plan designs on how not to block critical experience interactions such as swipe, while allowing greater flexibility with the consent design system.

Given that work hasn't been planned and designs don't exist yet, I would expect this to take a few months at least.  Perhaps there is some amount of overlap (from a user experience design perspective) with the blocking experience for paywalls currently being designed.

/cc @hongcatlover

**Comment by AramZS** [comment]
Still wondering if this could ever happen?

**Comment by newmuis** [comment]
We have been considering this approach, but have still not devised or been made aware of any approach that would allow us to enable this consent use case without jeopardizing the overall user experience.  There are strong concerns that people would misuse this to enable other experiences, such as pre-roll advertisements, that are normally disallowed in story experiences.

Do you need the stories to be AMP valid?

**Comment by AramZS** [comment]
@newmuis This is a major issue. I mean, to be clear this means that AMP Stories is out of compliance for the EU. Like... entirely out of compliance if you run an ad in the story. It makes the entire AMP Stories approach impossible to monetize in the EU and, considering that privacy laws are only becoming more common, potentially a lot of other places. If the user has made a choice to opt out and we can't synchronize that choice in AMP Stories there are *laws* being *broken*. Like, at this moment, **right now**. 

AMP Stories has basically gotten a weird pass on this because it is an evolving tech deployed in limited contexts and relatively new, but this issue has now been open for almost an entire year and that really doesn't hold water anymore. I can't speak to how the details of what is or is not allowed inside an `amp-consent` tag might enable other things to run, but if there's any inclination to keep this project alive then being compliant with international laws seems like it would be a requirement? IMO it's worth prioritizing how to address these concerns in a way that maintains your experience. 

Because we don't have this capability, the conversation I'm going to have to have with our internal team that develops AMP Stories is going to involve the question of if it is worthwhile to continue to use or maintain our AMP Stories at all anymore or if we have to build some sort of additional solution in front of how we serve these. That's the level of this issue. If the stories are not AMP valid than what's the point of using this technology at all? Might as well throw it out and use something that we have more control over the consent flows without having to hack in a bunch of stuff that constantly throws errors, right?

**Comment by newmuis** [comment]
Can you elaborate on the compliance issue?  Our understanding has been that this is a user experience issue, in that the current experience is that users would be prompted on one page, then prompted again on another, but each page correctly honors the user's response.  Thus, this is not the ideal user experience, and we'd like to seek solutions here, but we've not been aware of a compliance angle to this issue.

**Comment by AramZS** [comment]
It is legally required in a number of situations that if you have a record of a users' consent and therefore know their consent preference you are not permitted to re-prompt that user for consent again (especially if they have *declined* to consent to tracking) in any context without meeting some particularly high bars that this does not. 

We have user accounts, we know these users' status, including those that did not consent to tracking, and we would like to make sure that consent is synchronized across our entire domain, including AMP Stories, so that we do not fall out of compliance.

**Comment by newmuis** [comment]
Can you verify the question above regarding AMP validity?

The domain is quite literally the problem -- the consent is not synchronized across the web site and the stories because they are served from different domains and browser restrictions prevent the storage from being read across domains.  If you do not require the content to be valid AMP, you can serve the story from the same domain as other content and this would be a problem that is solved much more easily.

**Comment by AramZS** [comment]
We are currently serving the story on our domain, but we cannot trigger the consent systems in AMP (like hiding an ad for example) with the users' consent settings if we can't query our consent API in the way we do with standard AMP. All the consent functionality in AMP is locked off by the AMP consent component and that component can't manage it with the data from our systems. And if we can't serve valid AMP, that's *bad*. The system should not be invalidated by following the legal requirements for consent. This is solved in standard AMP in a really satisfactory way, it *isn't* in AMP Stories.

I remain very unclear why the existing consent flow cannot be adopted into AMP Stories.

**Comment by newmuis** [comment]
The existing consent flow cannot be adopted into stories because it is not limited to consent.  The existing consent flow has an abuse vector which would allow for publishers to create valid AMP stories that iframe other pages that do not meet the same user experience principles used for stories.  It would invalidate the guarantee that is currently provided to consumption platforms that valid AMP story content adheres to the story user experience model; one could publish valid AMP story content that is effectively not treated as a story.

This is not an issue for AMP websites, as they do not enforce any such user experience principles.  Using the existing AMP solution for this will definitely be a non-starter; they are different products with different user needs, and as such, they require different solutions.

To be clear, AMP stories _do_ still use the AMP consent component, and are able to interface with other AMP pages.  What is not currently supported is the ability to do so across domain boundaries.  It is possible for amp-consent to make requests, but it may be difficult for publishers to reconcile this server-side, without sufficient tracking information.

We're open to engineering design discussions for how we might reconcile these requirements, but we are not currently aware of a solution here.  The closest engineering design we've come to that meets the product and user experience requirements but also allows cross-domain consent and custom CMP UI would be:

- The framework can allow a hidden iframe that does not render any UI, but can be served off of the publisher's canonical origin, to communicate cookie state to the story
- CMPs can follow the [process for significant changes](https://github.com/ampproject/amphtml/blob/main/docs/contributing-code.md#process-for-significant-changes) to create a bespoke `amp-story` implementation of their user interface
- Once available, publishers can use their CMP's version of the component

I think this can work at a high-level, but has the following caveats:

- Requires a heavy lift from CMPs, and low coverage for publishers (as every individual CMP would need to engage)
- Somewhat fragile implementation, as I believe this would break as browsers increase restrictions on third-party cookies (which I believe is also true of the [solution](https://community.cookiepro.com/s/article/UUID-94cf9b13-1998-d989-a5e0-d01a73a2e0a7?language=en_US) provided upthread)

**Comment by AramZS** [comment]
I think they'd only need to have an iframe without a custom user interface? This is getting deeper into the components than I usually do and I'm only operating from experience in regard to our previous setup. We could have the hidden iframe approach and postMessage the user's state to the existing UI perhaps? 

I agree that this flow is threatened by the changes to 3p cookies, but... as you said... so is the existing AMP flow? Getting it working under the current model established in standard AMP seems like a fairly low bar to accomplish before 3p is dead. I understand your argument that the way it currently works might allow other flows that would break Stories, but you managed to alter the methodology and restrictions of other elements on behalf of Stories, it can't be impossible to do so with `<amp-consent>`? 

Also, since you mentioned it: should I take this post as an indicator that standard AMP's consent flow is expected to break, with no planned solution, as well?

**Comment by newmuis** [comment]
To be clear, it is not AMP's consent flow that will break, it is CookiePro by OneTrust's implementation that will break, since it is based on cookies.  They could alter this guidance to not depend on cookies, the implementation of which depends on the use cases they hope to enable, which I am not fully aware of.

Regarding the communication with the 3p page, perhaps there are actually two cases and we've conflated them:

1. Cases where publishers only need the UI provided by the existing `amp-story-consent`, but need to link the data to an external source.  For this, we may be able to use an invisible iframe, and postMessage its state to the story
2. Cases where publishers need custom UI.  For this, CMPs would need to implement such UI as an AMP extension within the repository directly.

Without any UI-facing change, we could proceed with (1), and we would need to go through the [process for significant changes](https://github.com/ampproject/amphtml/blob/main/docs/contributing-code.md#process-for-significant-changes).

I think there is enough information in this thread to reformat into a lightweight [intent to implement](https://github.com/ampproject/amphtml/issues/new?assignees=&labels=INTENT+TO+IMPLEMENT&template=intent-to-implement.yml) that concretely proposes the already with the invisible iframe.  I can act as your [guide](https://github.com/ampproject/amphtml/blob/main/docs/contributing-code.md#find-a-guide), so then the next step would be to [sign up for an AMP design review](https://github.com/ampproject/amphtml/blob/main/docs/design-reviews.md#bringing-your-design-to-a-design-review).

Just to note, again, this would only solve the case of using the existing UI coupled to data from an external domain, but not the case of using a CMP-defined UI, which I believe is what @vyoungnyc and @cmaurersp are proposing.  For that, we would still need to separately explore CMPs submitting their own components, which would require separate intents to implements and reviews.

**Comment by AramZS** [comment]
Yes, I don't care about the UI, sorry if that was unclear, all I care about is a consistent consent state for users that can be synced with a third party like OneTrust. I don't care about the UI for it and I doubt anyone else does either as long as you can get the consent synced. UI change here is a much lesser interest. 

And to be clear, the consent flow is already broken by not doing this. It is non-compliant because it cannot read previous consent states as set by the containing website, or a third party that handles consent for them. 

Just to double check the process here, it seems like what you are proposing is that we write the code and submit it to review?

**Comment by newmuis** [comment]
The process would be:

- [x] Interested parties should propose a solution (I believe this has happened in this thread: an invisible iframe that communicates state via postMessage)
- [ ] This solution should be documented in an [intent to implement](https://github.com/ampproject/amphtml/issues/new?assignees=&labels=INTENT+TO+IMPLEMENT&template=intent-to-implement.yml)
- [ ] That intent to implement should be reviewed in an [AMP design review](https://github.com/ampproject/amphtml/blob/main/docs/design-reviews.md#bringing-your-design-to-a-design-review)
- [ ] Feedback from design review should be addressed (and re-reviewed in a subsequent AMP design review, if needed)

At this point, we should know what code needs to be written, which is a prerequisite for understanding who will write it, and when.
